clippy_utils/src/higher.rs

```diff
@@ -470,6 +470,42 @@ pub fn extract_assert_macro_args<'tcx>(e: &'tcx Expr<'tcx>) -> Option<Vec<&'tcx
     None
 }
 
+/// A parsed `matches!` expansion
+#[derive(Debug)]
+pub struct MatchesExpn<'tcx> {
+    /// Span of `matches!(..)`
+    pub call_site: Span,
+    /// Second parameter of `matches!`
+    pub arm: &'tcx Arm<'tcx>,
+    /// First parameter of `matches!`
+    pub exp: &'tcx Expr<'tcx>,
+}
+
+impl MatchesExpn<'tcx> {
```

Comment:
Ah, we don't really want to keep doing this: #7843

Is this only required to generate the suggestion? If so, I'd prefer to just have a non-auto-applicable suggestion in the `matches` case with `Applicability::HasPlaceholder`

Reply:
Parameters of `matches!` macro are needed to detect if this pattern is equatable.

Reply:
Ah, that's unfortunate. We really have to figure out how to do this without relying on the exact expansion of the macro...